posts/onboarding.md

```diff
@@ -0,0 +1,525 @@
+
+Python is not designed for maximum processing speed/performance, so SQLGlot maintains an equivalent [Rust version of the tokenizer](https://tobikodata.com/sqlglot-jumps-on-the-rust-bandwagon.html) in `sqlglotrs/tokenizer.rs` for improved performance.
+
+> [!IMPORTANT]
+> Changes in the tokenization logic must be reflected in both the Python and the Rust tokenizer. The goal is for the two implementations to be similar so that there is less cognitive effort to port code between them.
+
+This example demonstrates the results of using the Tokenizer to tokenize the SQL query `SELECT b FROM table WHERE c = 1`:
+
+```python
+from sqlglot import tokenize
+
+tokens = tokenize("SELECT b FROM table WHERE c = 1")
+for token in tokens:
+    print(token)
+
+# <Token token_type: <TokenType.SELECT: 'SELECT'>, text: 'SELECT', line: 1, col: 6, start: 0, end: 5, comments: []>
+# <Token token_type: <TokenType.VAR: 'VAR'>, text: 'b', line: 1, col: 8, start: 7, end: 7, comments: []>
+# <Token token_type: <TokenType.FROM: 'FROM'>, text: 'FROM', line: 1, col: 13, start: 9, end: 12, comments: []>
+# <Token token_type: <TokenType.TABLE: 'TABLE'>, text: 'table', line: 1, col: 19, start: 14, end: 18, comments: []>
+# <Token token_type: <TokenType.WHERE: 'WHERE'>, text: 'WHERE', line: 1, col: 25, start: 20, end: 24, comments: []>
+# <Token token_type: <TokenType.VAR: 'VAR'>, text: 'c', line: 1, col: 27, start: 26, end: 26, comments: []>
+# <Token token_type: <TokenType.EQ: 'EQ'>, text: '=', line: 1, col: 29, start: 28, end: 28, comments: []>
+# <Token token_type: <TokenType.NUMBER: 'NUMBER'>, text: '1', line: 1, col: 31, start: 30, end: 30, comments: []>
+```
+
+The Tokenizer processes each of the query’s eight symbols (words or operators like `=`) by looking up each one in its mapping of keyword strings to token type.
+
+Each token is generally represented by its `TokenType` and by its string value stored in the `text` field. Other metadata such as `line` or `col` are stored during this step and can help with error diagnostics and other tasks.
+
+SQLGlot’s `TokenType` enum provides an indirection layer between the text of each symbol and its type. For example, `!=` and `<>` are often used interchangeably to represent "not equals", so SQLGlot groups them together by mapping them both to `TokenType.NEQ`.
+
+The mapping of symbols to `TokenType` enum values is stored in the dictionary `Tokenizer.KEYWORDS`.
+
+
+
+## Parser
+
+The parser module takes the list of tokens generated by the tokenizer and builds an [abstract syntax tree (AST)](https://en.wikipedia.org/wiki/Abstract_syntax_tree) from them.
+
+The AST represents the meaning (“semantics”) of a SQL statement, loosely defined as the set of actions the statement should invoke. The commonality of semantics across databases/dialects is what allows transpilation to occur.
```

Comment:
```suggestion
The AST represents a SQL statement by storing its _meaningful_ constituent components. For example, to represent a `SELECT` query we can only store its projections, filters, aggregation expressions and so on, but the `SELECT`, `WHERE` and `GROUP BY` keywords don't need to be preserved, since they are implied.

In SQLGlot's case, the AST usually stores information that captures the _semantics_ of an expression, i.e. its meaning, which is what allows it to transpile SQL code between different dialects. A blog post that explains this idea at a high-level can [be found here](https://tobikodata.com/transpiling_sql1.html).
```

Reply:
This reads a bit off to me, the second paragraph seems disconnected and/or repeating the first, should we merge them?

Reply:
Yeah, I can see how this may be confusing -- I'm open to rephrasing it.

Imo we shouldn't conflate the _semantics_ of an expression with its grammatical structure, which is what an AST is primarily meant to represent.

SQLGlot's AST is special, because we inject additional information to it as opposed to what you'd get by a typical parser. For example, some `Func` nodes store a return type, and other nodes implicily store whether the arguments can be strings or datetime values (e.g. the `TsOrDs*` family).

That's what I usually think about when I see the term "semantics". Also, keep in mind that we don't really represent the semantics accurately in many cases, e.g. when expressions are type-sensitive.

Reply:
I see what you mean, I think the confusing part is that I'd read `The AST represents ...` as basically `The [SQLGLot] AST represents ...` which can be conflicting with the 2nd paragraph. From what I understand your intention is to distinguish a typical AST (i.e storing just the grammatical structure) produced from a general purpose SQL parser to the "enhanced" version that is produced by SQLGlot. 

If this is the case, I think your 2nd explanation conveys the message better

Reply:
What about:

> An AST generally stores the _meaningful_ constituent components of a statement. For example, to represent a `SELECT` query we can only store its projections, filters, aggregation expressions and so on, but the `SELECT`, `WHERE` and `GROUP BY` keywords don't need to be preserved, since they are implied.

> In SQLGlot's case, the AST usually stores additional information that captures the _semantics_ of an expression, i.e. its meaning, which is what allows it to transpile SQL code between different dialects. A blog post that explains this idea at a high-level can [be found here](https://tobikodata.com/transpiling_sql1.html).